torch/jit/__init__.py

```diff
@@ -727,6 +727,26 @@ def _is_new_style_class(cls):
         return ('__dict__' in dir(cls) or hasattr(cls, '__slots__'))
 
 
+# Retrieves a fully-qualified name (module hierarchy + classname) for a given obj.
+def _qualified_name(obj):
```

Comment:
I believe there should also be a check: 
```getattr(lookup_module(module_name), name) is obj```

Otherwise bad things can happen if someone tries to just set __module__. This is how the pickler works as well.

Reply:
This unfortunately doesn't work with nested classes (or classes defined in functions), since python pre-3.3 doesn't have a way of figuring out the nested name within a module.

So adding this check makes writing the tests very tricky, since we have to un-nest everything. I'd prefer to make that change in a follow-up since this PR is big already.